### DISCUSSION TITLE
Creating a custom MuJoCo plugin

### DISCUSSION THREAD

**Discussion by gavin-hyl** [google-deepmind/mujoco#2045] [Asking for Help] [answered]
### Intro

Hi!

I am a undergrad at Caltech.

### My setup

Windows latest version, latest MJX/Mujoco python API

### My question

I am currently trying to simulate a robot that has motors on its joints. The motors have built-in PD controllers with a feed-forward term. How can I achieve this functionality with a single custom MuJoCo plugin?

I tried using the built-in PID extension as a template to write the PDFF controller, but even though my code compiles, I could not get MuJoCo to find and use it when I call it in the XML file. Right now, I compiled my code to a `libpdff_plugin.so` file, and placed it into the same folder as the XML file. 

Please review the following stack overflow question:
https://stackoverflow.com/questions/78917806/creating-a-custom-mujoco-extension

### Minimal model and/or code that explain my question

Here's the python script that tries to load the XML:

```
import mujoco
import os

plugin_path = "libpdff_plugin.so"
os.environ["LD_PRELOAD"] = plugin_path

XML = "example.xml"
model = mujoco.MjModel.from_xml_path(XML)

print(model)
```

and here is the XML itself:

```
<mujoco>
  <extension>
    <plugin plugin="pdff">
      <instance name="pdff_instance">
        <config key="kp" value="40.0"/>
        <config key="kd" value="4"/>
        <config key="ff_setpoint" value="0"/>
        <config key="ctrl_max" value="1"/>
      </instance>
    </plugin>
  </extension>

  <worldbody>
    <body>
      <joint name="j" type="slide" axis="0 0 1" />
      <geom type="sphere" size="0.01" mass="1"/>
    </body>
  </worldbody>

  <actuator>
    <plugin joint="j" plugin="pdff" instance="pdff_instance"/>
  </actuator>
</mujoco>
```

The script gives the below error:

```
Traceback (most recent call last):
  File (the parent directory)/test.py", line 8, in <module>
    model = mujoco.MjModel.from_xml_path(XML)
ValueError: XML Error: unknown plugin 'pdff'
Element 'plugin', line 3
```

### Confirmations

- [X] I searched the [latest documentation](https://mujoco.readthedocs.io/en/latest/overview.html) thoroughly before posting.
- [X] I searched previous [Issues](https://github.com/google-deepmind/mujoco/issues) and [Discussions](https://github.com/google-deepmind/mujoco/discussions), I am certain this has not been raised before.

**Comment by kevinzakka** [comment, marked as answer]
I assume you're using the feedforward term for gravcomp? In that case, I think you can achieve the functionality you want without a plugin, using a combination of `motor` actuators and `qfrc_applied`.

```python
def gravcomp(
    model: mujoco.MjModel,
    data: mujoco.MjData,
    body_ids: np.ndarray,
) -> np.ndarray:
    """Compute force to counteract gravity for the given bodies."""
    jac = np.empty((3, model.nv))
    qfrc_gravcomp = np.zeros((model.nv))
    for i in body_ids:
        body_weight = model.opt.gravity * model.body(i).mass
        mujoco.mj_jac(model, data, jac, None, data.body(i).xipos, i)
        q_weight = jac.T @ body_weight
        qfrc_gravcomp -= q_weight
    return qfrc_gravcomp


def pd_controller(
    model: mujoco.MjModel,
    data: mujoco.MjData,
    qpos_desired,
    Kp: np.ndarray,
    Kd: np.ndarray,
) -> None:
    """u = Kp * (q_d - q) - Kd * qdot"""
    tau = Kp * (qpos_desired - data.qpos) - Kd * data.qvel
    np.clip(tau, *model.actuator_ctrlrange.T, out=data.ctrl)


def joint_stiffness_controller(
    model: mujoco.MjModel,
    data: mujoco.MjData,
    qpos_desired,
    Kp: np.ndarray,
    Kd: np.ndarray,
    body_ids: Sequence[int],
) -> None:
    """PD controller + gravity compensation.

    u = -tau_gravity + Kp * (q_d - q) -Kd * qdot
    """
    tau_gravcomp = gravcomp(model, data, body_ids)
    data.qfrc_applied = tau_gravcomp
    tau = Kp * (qpos_desired - data.qpos) - Kd * data.qvel
    np.clip(tau, *model.actuator_ctrlrange.T, out=data.ctrl)
```

You can use the above controller helper functions in the following control loop:

```python
# MuJoCo model and data pair.
model = ...
data = ...
# List of bodies to apply gravity compensation to.
body_ids: list[int]
# PD controller gains.
Kp: np.ndarray
Kd: np.ndarray

while True:
  qpos_target: np.ndarray  # Assume some higher level logic is providing this.
  joint_stiffness_controller(model, data, qpos_target, Kp, Kd, body_ids)
  mujoco.mj_step(model, data)
```

**Reply by gavin-hyl** [reply]
Thanks! That was exactly what I was looking for.
